### PR TITLE
add unit test for fused_rmsnorm, add testing_utils

### PR DESCRIPTION
**Pull request by lessw2020**
This PR adds a basic numerical accuracy unit test to compare the fused_rmsnorm recently added vs the PyTorch RMSNorm.

1 - defaults to a simple run of bs=8, tensor of size (8192).  Compares using torch.float32 to avoid rounding issues during test. 

2 - verifies accuracy to atol=1e-7

3 - verifies accuracy on both forward pass and then backward pass using a randomized, synthetic gradient

3 - adds testing_utils which has a wrapper for asserting with allclose, setting seed, gpu_test. gpu_test is needed to ensure cpu tests pass, as Triton kernels do not run on cpu. 

Screenshot 2024-04-07 at 10 52 10 PM

### CODE REVIEW

File: test/testing_utils.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# This software may be used and distributed according to the terms of the Llama 2 Community License Agreement.
+
+import time
+from typing import Any, Optional
+
+import pytest
+import torch
+
+
+def perf_timer(func):
```
Comment (awgu) [inline test/testing_utils.py]: This looks unused in this PR?

File: test/test_fused_rmsnorm.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+import sys
+
+import pytest
+import torch
+import torch.nn as nn
+from torch import Tensor
+
+sys.path.append("..")
+from torchtrain.models.norms import FusedRMSNorm  # fused_rms_norm_fn as triton_rmsnorm
```
Comment (awgu) [inline test/test_fused_rmsnorm.py]: nit: if not using the `fused_rms_norm_fn` import anymore
```suggestion
from torchtrain.models.norms import FusedRMSNorm
```

File: test/testing_utils.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        actual,
+        expected,
+        rtol=rtol,
+        atol=atol,
+        check_device=check_device,
+        msg=f"actual: {actual}, expected: {expected}",
+    )
+
+
+def set_rng_seed(seed):
```
Comment (awgu) [inline test/testing_utils.py:27]: `set_rng_seed` seems like an extra layer of indirection that might not be needed (for now)?

File: test/testing_utils.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    """Sets the seed for pytorch random number generators"""
+    torch.manual_seed(seed)
+
+
+def gpu_test(gpu_count: int = 1):
```
Comment (awgu) [inline test/testing_utils.py:32]: This looks unused.

File: test/test_fused_rmsnorm.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    """Root Mean Square Layer Normalization
+    as proposed in: https://arxiv.org/abs/1910.07467
+
+    Calcs are done in fp32.
+
+    original impl: https://github.com/facebookresearch/llama/blob/main/llama/model.py
+
+    Args:
+        dim(int):  model size
+        eps(float): epsilon
```
Comment (awgu) [inline test/test_fused_rmsnorm.py]: suggestion:
```suggestion
```

File: test/test_fused_rmsnorm.py (hunk 2 above)
Comment (lessw2020) [reply test/test_fused_rmsnorm.py]: removed.

File: test/testing_utils.py (hunk 4 above)
Comment (lessw2020) [reply test/testing_utils.py:32]: it is atm, but I added it b/c it seems we have cpu tests and gpu tests and Triton tests will fail on cpu tests...I'm not clear how our cpu vs gpu tests are being determined (since I would expect most torchtrain functionality would need a gpu), but that's why I included it.

Comment (lessw2020) [reply test/testing_utils.py:32]: ok cpu tests failing (apparently it's just a global grab of all tests), so putting this decorator to use :)

Comment (lessw2020) [reply test/testing_utils.py:32]: cpu tests now passing, so will leave this gpu decorator in which resolves this issue (now using).

File: test/testing_utils.py (hunk 3 above)
Comment (lessw2020) [reply test/testing_utils.py:27]: I'd like to leave it in as it will almost certainly have it's place in many tests. (you are right that technically the current test I use it in is not vital to have it seeded).

File: test/testing_utils.py (hunk 1 above)
Comment (lessw2020) [reply test/testing_utils.py]: it is, but I was using it earlier to verify perf delta on fused vs PT RMS.  I have removed.

File: test/test_fused_rmsnorm.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+from torchtrain.test.testing_utils import assert_expected, gpu_test, set_rng_seed
+
+
+@pytest.fixture(autouse=True)
+def set_seed():
+    set_rng_seed(2023)
+
+
+class TorchRMSNorm(nn.Module):
```
Comment (tianyu-l) [inline test/test_fused_rmsnorm.py:23]: can we reuse the implementation in torchtrain/models/norms.py?

Comment (lessw2020) [reply test/test_fused_rmsnorm.py:23]: I thought about this but my concern was we want to ensure the 'correct' answer in the unit test is always correct and therefore I made a separate impl here.  
If someone (accidentally) modifies the RMSNorm in torchtrain, which is where I would predict it would be much more likely to happen, then it would currently show only that fused was incorrect, when the real issue is that the torchtrain RMSNorm is no longer correct.
